### PR TITLE
DiscoveryFeed: add tale to desktop

### PR DESCRIPTION
**Pull request by erikos**
https://phabricator.endlessm.com/T16635

### CODE REVIEW

File: js/js-resources.gresource.xml [hunk 1]
```diff
@@ -112,12 +112,13 @@
     <file>ui/components/appStore.js</file>
     <file>ui/components/autorunManager.js</file>
     <file>ui/components/automountManager.js</file>
-    <file>ui/components/keyring.js</file>
+    <file>ui/components/codingManager.js</file>
     <file>ui/components/codingGameService.js</file>
+    <file>ui/components/discoveryFeed.js</file>
+    <file>ui/components/keyring.js</file>
```
Comment (smspillaz) [inline js/js-resources.gresource.xml]: Can probably revert the change here which moves `keyring.js` around

File: js/ui/viewSelector.js [hunk 2]
```diff
@@ -53,16 +53,18 @@ const ViewsDisplayLayout = new Lang.Class({
     Name: 'ViewsDisplayLayout',
     Extends: Clutter.BinLayout,
 
-    _init: function(entry, allViewActor, searchResultsActor) {
+    _init: function(entry, gc, allViewActor, searchResultsActor) {
         this.parent();
 
         this._entry = entry;
+        this._gc = gc;
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: Probably rename this to `discoveryFeed` given that we are not using the old codename anymore.

File: js/ui/viewSelector.js [hunk 3]
```diff
@@ -53,16 +53,18 @@ const ViewsDisplayLayout = new Lang.Class({
...
         this._allViewActor = allViewActor;
         this._searchResultsActor = searchResultsActor;
 
         this._entry.connect('style-changed', Lang.bind(this, this._onStyleChanged));
         this._allViewActor.connect('style-changed', Lang.bind(this, this._onStyleChanged));
 
+        this._heightAboveGC = 0;
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: It seems like this is always zero. Can it be made a constant instead of an instance variable?

File: js/ui/viewSelector.js [hunk 4]
```diff
@@ -284,7 +303,19 @@ const ViewsDisplay = new Lang.Class({
...
+        let icon = new St.Icon({ icon_name: 'edit-find-symbolic',
+                                style_class: 'discovery-feed-icon',
+                                icon_size: 16,
+                                track_hover: true });
+        this.gc = new St.Button({ style_class: 'discovery-feed',
+                                  x_align: Clutter.ActorAlign.CENTER,
+                                  y_align: Clutter.ActorAlign.CENTER });
+        this.gc.set_child(icon);
+        this.gc.connect('clicked', Lang.bind(this, function() {
+          Main.discoveryFeed.show(0);
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: Should we be calling this with an actual timestamp instead of zero?

File: js/ui/viewSelector.js [hunk 5]
```diff
@@ -284,7 +303,19 @@ const ViewsDisplay = new Lang.Class({
...
+        }));
+
+        this.actor = new ViewsDisplayContainer(this.entry, this.gc, this._allView, this._searchResults);
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: What is the difference between `this.gc` which is passed to `ViewsDisplayContainer` and `gc` passed to `ViewsDisplayLayout` below?

File: js/ui/viewSelector.js [hunk 6]
```diff
@@ -132,6 +139,16 @@ const ViewsDisplayLayout = new Lang.Class({
...
 
+        let gcHeight = this._gc.get_preferred_height(availWidth)[1];
+
+        let gcBox = box.copy();
+        let x1 = (availWidth - this._gc.get_width()) * 0.5;
+        gcBox.y1 = this._heightAboveGC;
+        gcBox.y2 = gcBox.y1 + gcHeight;
+        gcBox.x1 = x1;
+        gcBox.x2 = x1 + this._gc.get_width();
+        this._gc.allocate(gcBox, flags);
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: While this is consistent with the existing code, my preference here would be to create a subclass of `St.Button` for the discovery feed handle and encapsulate as much of its code as possible inside of instance methods. We would only need to pass `availWidth` and `this._heightAboveGC` to it.

File: js/ui/viewSelector.js (hunk 4 above)
Comment (smspillaz) [inline js/ui/viewSelector.js]: See comment above about encapsulating a lot of this logic into its own subclass of `St.Button`

**Comment by smspillaz** [comment]
Tested, everything seems to work fine with the gsetting and using the corresponding branch on e-d-f

File: js/ui/components/discoveryFeed.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+const Lang = imports.lang;
+
+const Main = imports.ui.main;
+const SideComponent = imports.ui.sideComponent;
+
+const DISCOVERY_FEED_NAME = 'com.endlessm.DiscoveryFeed';
+const DISCOVERY_FEED_PATH = '/com/endlessm/DiscoveryFeed';
+
+const DiscoveryFeedIface = '<node> \
+<interface name="' + DISCOVERY_FEED_NAME + '"> \
```
Comment (smspillaz) [inline js/ui/components/discoveryFeed.js]: For consistency here, I'd add a `const DISCOVERY_FEED_IFACE` that explicitly specifies the interface name. Some of the trouble I was debugging last night was because of confusion between the dbus name and the interface

File: js/ui/viewSelector.js [hunk 8]
```diff
@@ -416,6 +476,8 @@ const ViewsClone = new Lang.Class({
                                                           enabled: false });
         iconGridClone.add_effect(this._saturation);
 
+        if (discoveryFeed !== null)
+            this.add_child(discoveryFeed);
```
Comment (smspillaz) [inline js/ui/viewSelector.js]: Instead of checking whether discoveryFeed is null here, you can probably just add the child in the parent if condition

Comment (erikos) [inline js/ui/viewSelector.js]: I actually tried that, but it has to happen after the constructing of the ViewsDisplayLayout.
